### PR TITLE
HHH-17384 OneToMany association with @NotFound results in SQL with different JOIN-type for SELECT (LEFT JOIN) and COUNT (JOIN)

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-17384

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java [hunk 1]
```diff
@@ -3774,18 +3784,28 @@ private TableGroup createTableGroup(TableGroup parentTableGroup, SqmPath<?> join
 				querySpec.getFromClause().addRoot( tableGroup );
 			}
 			else {
+				final SqlAstJoinType sqlAstJoinType;
+				if ( currentClauseStack.getCurrent() != Clause.SELECT
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java:3788]: Apparently this is not the case, if I remove the `currentClauseStack.getCurrent() != Clause.SELECT` condition I get test failures. In particular: 
* `NotFoundExceptionLogicalOneToOneTest#testQueryAssociationSelection`
* `NotFoundExceptionManyToOneTest#testQueryAssociationSelection`
* `NotFoundIgnoreManyToOneTest#testQueryAssociationSelection`
* `NotFoundIgnoreOneToOneTest#testQueryAssociationSelection`

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java [hunk 2]
```diff
@@ -3774,18 +3784,28 @@ private TableGroup createTableGroup(TableGroup parentTableGroup, SqmPath<?> join
...
+						&& getInferredValueMapping() == null
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java:3789]: The reason to use `LEFT` join only when the inferred mapping is null is we need it just for nullness predicates, otherwise we always want `INNER` joins for i.e. comparison operations.

WDYT of something like:  
```
// If we have an inferred value mapping we are using this path "against" another, 
// and we always want INNER Join semantics in that case
```

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java [hunk 3]
```diff
@@ -3774,18 +3784,28 @@ private TableGroup createTableGroup(TableGroup parentTableGroup, SqmPath<?> join
...
+						&& joinProducer instanceof ToOneAttributeMapping
+						&& ( (ToOneAttributeMapping) joinProducer ).hasNotFoundAction() ) {
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java:3791]: We don't always call `upgradeToInnerJoinIfNeeded` after creating the table group, there are more code paths that call `BaseSqmToSqlAstConverter#createTableGroup`, for example when the assocation path is not de-referenced like in our case (e.g. `e.association is null).

Also, upgrading the join type to `INNER` after requires processing to find the parent `TableGroup` and the related `TableGroupJoin`, while in `createTableGroup` we already have those objects so creating the correct type early would be more efficient.

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java [hunk 4]
```diff
@@ -1107,20 +1108,27 @@ protected Predicate determineWhereClauseRestrictionWithJoinEmulation(AbstractUpd
 					querySpec.applyPredicate( tableReferenceJoin.getPredicate() );
 				}
 				for ( TableGroupJoin tableGroupJoin : root.getTableGroupJoins() ) {
-					assert tableGroupJoin.getJoinType() == SqlAstJoinType.INNER;
-					querySpec.getFromClause().addRoot( tableGroupJoin.getJoinedGroup() );
-					querySpec.applyPredicate( tableGroupJoin.getPredicate() );
+					if ( tableGroupJoin.getJoinType() == SqlAstJoinType.INNER ) {
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:1111]: By avoiding the nested subquery you mean removing the left join special handling in `visitNullnessPredicate`, right?

In any case, I've tried implementing this but sadly, at least on H2, I get the error `Column "PERSON.ACCOUNT_ID" not found` for the query:
```sql
    delete 
    from
        Person 
    where
        exists(select
            1 
        from
            ACCOUNT_TABLE la1_0 
        left join
            ACCOUNT_TABLE a1_0 
                on a1_0.id=Person.account_id 
        where
            la1_0.id=Person.lazyAccount_id 
            and a1_0.id is null 
            and la1_0.code<>'aaa')
```
I guess you can't left join using a column from the target of a mutation statement.

[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:1111]: At least on PostgreSQL and MySQL this works and produces a better execution plan. Now, if it doesn't work on H2, then we'd indeed have to use the double nested subquery approach for H2, but I would prefer if we have a code path for "sane" databases to avoid double nesting since that comes with a bunch of other problems.

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java:3791]: Ok, but the point about abstracting the method to handle `@OneToOne(mappedBy = "...")` the same way is still valid. Could you please add a test for that and ensure the `is null` predicate works for such associations as well?

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java:3788]: Ok, this is because the upgrade to inner join is only done if a table group already exists for the path. If the table group is created as part of resolving the path, then `prepareForSelection` assumes that the join already uses `inner`, which is kind of wrong now since we changed the default join type.

A HQL query like `select e.notFoundAssociation from MyEntity e` should produce a query like this:

```sql
select *
from my_entity e
join association_table a on e.fk = a.pk
```

I guess with the current state, it will produce a left join now? To fix that, you'd need to call `upgradeToInnerJoinIfNeeded` in the `if ( tableGroup == null ) {` cas in `prepareForSelection`.

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java:3789]: The inferred mapping can be null in the following potentially problematic cases:

1. top-level select item
2. group by expression
3. tuple expression
4. select item of exists subquery

Case 1. and 4. should be fixed by changing `prepareForSelection` like I suggested above. Then you also wouldn't need to care about the SELECT clause in here. No idea how to fix 2. and 3. though. Problematic means that e.g. `from MyEntity e group by e.notFoundAssociation` would produce a `left join` for the association, just like a `from MyEntity e where (e.id,e.notFoundAssociation)=(1,:param)` would.

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java]: I have added an implementation which can handle both the virtual table group + left join and the "old" roots implementation based on a new `supportsJoinInMutationStatementSubquery()` flag. H2 and Derby seem to be the only databases that complain about the subquery's join.

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java]: I have implemented a solution which passes an `allowLeftJoins` flag only when visiting an `is null` predicate. This allows searching for an existing join of the correct type (`left` for not found to-one associations) and eventually upgrade it if the context requires `inner` joins.

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java]: I have added a test for the `mappedBy` case and handled that in the `ToOneAttributeMapping#hasNotFound()` method.

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java]: See my previous comment for how I've handled join "upgrades" to `inner`.

File: hibernate-core/src/main/java/org/hibernate/dialect/DerbySqlAstTranslator.java [hunk 5]
```diff
@@ -261,6 +261,11 @@ private boolean supportsOffsetFetchClause() {
 		return true;
 	}
 
+	@Override
+	protected boolean supportsJoinInMutationStatementSubquery() {
+		return false;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/DerbySqlAstTranslator.java:266]: Might be worth adding that to the legacy variant as well.

File: hibernate-core/src/main/java/org/hibernate/dialect/H2SqlAstTranslator.java [hunk 6]
```diff
@@ -281,4 +281,9 @@ private boolean supportsOffsetFetchClausePercentWithTies() {
 		// Introduction of PERCENT support https://github.com/h2database/h2database/commit/f45913302e5f6ad149155a73763c0c59d8205849
 		return true;
 	}
+
+	@Override
+	protected boolean supportsJoinInMutationStatementSubquery() {
+		return false;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/H2SqlAstTranslator.java:287]: Might be worth adding that to the legacy variant as well.

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java [hunk 7]
```diff
@@ -2264,7 +2264,14 @@ public boolean isIgnoreNotFound(){
...
 
 	public boolean hasNotFoundAction() {
-		return notFoundAction != null;
+		if ( notFoundAction != null ) {
+			return true;
+		}
+		else if ( referencedPropertyName != null && cardinality != Cardinality.MANY_TO_ONE ) {
+			final AttributeMapping mappedByAttribute = entityMappingType.findAttributeMapping( referencedPropertyName );
+			return mappedByAttribute instanceof ToOneAttributeMapping && ( (ToOneAttributeMapping) mappedByAttribute ).hasNotFoundAction();
+		}
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java:2273]: Maybe I was a bit unclear on this, but what I meant is that a one-to-one association which uses `mappedBy` should behave the same way as if `@NotFound` is annotated. You don't have to check if the inverse attribute has a not found action, this is irrelevant.

File: hibernate-core/src/test/java/org/hibernate/orm/test/notfound/IsNullAndNotFoundMappedByTest.java [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+	public static class Account {
+		@Id
+		private Integer id;
+
+		private String code;
+
+		private Double amount;
+
+		@OneToOne
+		@NotFound( action = NotFoundAction.IGNORE )
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/notfound/IsNullAndNotFoundMappedByTest.java:318]: This should not be necessary for the test.

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java [hunk 9]
```diff
@@ -1092,6 +1095,59 @@ protected Predicate determineWhereClauseRestrictionWithJoinEmulation(AbstractUpd
...
+					if ( !virtualTableGroup.isSet() ) {
+						virtualTableGroup.set( new StandardVirtualTableGroup(
+								joinedGroup.getNavigablePath(),
+								joinedGroup.getModelPart(),
+								joinedGroup,
+								joinedGroup.isFetched()
+						) );
+					}
+					virtualTableGroup.get().addTableGroupJoin( tableGroupJoin );
+					querySpec.applyPredicate( tableGroupJoin.getPredicate() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:1116]: Shouldn't this rather be the following?

```suggestion
					final FromClause fromClause = querySpec.getFromClause();
					if ( fromClause.getRoots().isEmpty() ) {
						final TableGroup joinedGroup = tableGroupJoin.getJoinedGroup();
						fromClause.addRoot( new StandardVirtualTableGroup(
								joinedGroup.getNavigablePath(),
								joinedGroup.getModelPart(),
								joinedGroup,
								joinedGroup.isFetched()
						) );
						querySpec.applyPredicate( tableGroupJoin.getPredicate() );
					}
					else {
						fromClause.getRoots().get( 0 ).addTableGroupJoin( tableGroupJoin );
					}
```

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java [hunk 10]
```diff
@@ -1092,6 +1095,59 @@ protected Predicate determineWhereClauseRestrictionWithJoinEmulation(AbstractUpd
...
+		}
+
+		if ( virtualTableGroup != null && virtualTableGroup.isSet() ) {
+			final TableGroup firstInnerJoinedGroup = virtualTableGroup.get().getUnderlyingTableGroup();
+			collectedNonInnerJoins.forEach( firstInnerJoinedGroup::addTableGroupJoin );
+			querySpec.getFromClause().addRoot( virtualTableGroup.get() );
+		}
+		else if ( querySpec.getFromClause().getRoots().isEmpty() ) {
+			return statement.getRestriction();
+		}
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:1141]:
```suggestion
		if ( querySpec.getFromClause().getRoots().isEmpty() ) {
			return statement.getRestriction();
		}
		else if ( collectedNonInnerJoins != null ) {
			final TableGroup firstInnerJoinedGroup = querySpec.getFromClause().getRoots().get( 0 );
			for ( int i = 0; i < collectedNonInnerJoins.size(); i++ ) {
				firstInnerJoinedGroup.addTableGroupJoin( collectedNonInnerJoins.get( i ) );
			}
		}
```

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java [hunk 11]
```diff
@@ -3774,18 +3789,30 @@ private TableGroup createTableGroup(TableGroup parentTableGroup, SqmPath<?> join
...
 			else {
-				// Check if we can reuse a table group join of the parent
-				final TableGroup compatibleTableGroup = parentTableGroup.findCompatibleJoinedGroup(
-						joinProducer,
-						SqlAstJoinType.INNER
-				);
+				final TableGroup leftJoinedTableGroup;
+				final SqlAstJoinType sqlAstJoinType;
+				if ( joinProducer instanceof ToOneAttributeMapping && ( (ToOneAttributeMapping) joinProducer ).hasNotFoundAction() ) {
+					leftJoinedTableGroup = parentTableGroup.findCompatibleJoinedGroup(
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java:3795]: Might be worth splitting extracting the join directly i.e. return the `TableGroupJoin` from a new method `TableGroup#findCompatibleJoin` and let `findCompatibleJoinedGroup` just do `TableGroupJoin j = findCompatibleJoin(...); return j == null ? null : j.getJoinedGroup();`. Then you  wouldn't have to search for the join again later below.

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java (hunk 10 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:1141]: This won't work because we would be applying `collectedNonInnerJoins` on the `StandardVirtualTableGroup` we created earlier. This would cause them to get rendered as roots, instead of actual e.g. `left join`s. That's why I added the joins on the underlying table group in the first place.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java]: Altering existing parts of the SQL AST (e.g. `TableGroup`) is unfortunately also not going to work. The AST is cached, so any changes you do are racy from a concurrency perspective and what you're doing here is especially problematic because the transformation is not even idempotent.
Maybe you'll simply have to create a copy of the first table group as `StandardTableGroup` instead of a  `StandardVirtualTableGroup` then?

**Review by beikov — APPROVED** [review]
